src/uucore/src/lib/features/format/argument.rs

```diff
@@ -69,64 +73,86 @@ impl<'a> FormatArguments<'a> {
-                    let mut chars = s.trim_matches(|c| c == '"' || c == '\'').chars();
-                    if let Some(first_char) = chars.next() {
-                        if chars.clone().count() > 0 {
-                            // Emit a warning if there are additional characters
-                            let remaining: String = chars.collect();
-                            show_warning!(
-                                "{}: character(s) following character constant have been ignored",
-                                remaining
-                            );
-                        }
-                        return first_char as u64; // Use only the first character
-                    }
-                    return 0; // Empty quotes
-                }
-                extract_value(u64::extended_parse(s), s)
-            }
-            _ => 0,
+            Some(FormatArgument::UnsignedInt(n)) => Ok(*n),
+            Some(FormatArgument::Unparsed(os)) => Self::get_num::<u64>(os),
+            _ => Ok(0),
         }
     }
 
-    pub fn next_extended_big_decimal(&mut self, position: &ArgumentLocation) -> ExtendedBigDecimal {
+    pub fn next_extended_big_decimal(
+        &mut self,
+        position: &ArgumentLocation,
+    ) -> Result<ExtendedBigDecimal, FormatError> {
         match self.next_arg(position) {
-            Some(FormatArgument::Float(n)) => n.clone(),
-            Some(FormatArgument::Unparsed(s)) => {
-                extract_value(ExtendedBigDecimal::extended_parse(s), s)
+            Some(FormatArgument::Float(n)) => Ok(n.clone()),
+            Some(FormatArgument::Unparsed(os)) => Self::get_num::<ExtendedBigDecimal>(os),
+            _ => Ok(ExtendedBigDecimal::zero()),
+        }
+    }
+
+    fn get_num<T>(os: &OsStr) -> Result<T, FormatError>
+    where
+        T: ExtendedParser + From<u8> + From<u32> + Default,
+    {
+        let s = os_str_as_bytes(os)?;
+        // Check if the string begins with a quote, and is therefore a literal
+        if let Some((&first, bytes)) = s.split_first() {
+            if (first == b'"' || first == b'\'') && !bytes.is_empty() {
+                let (val, len) = if let Some(c) = bytes
+                    .utf8_chunks()
+                    .next()
+                    .expect("bytes should not be empty")
+                    .valid()
+                    .chars()
+                    .next()
+                {
+                    // Valid UTF-8 character, cast the codepoint to u32 then T
+                    // (largest unicode codepoint is only 3 bytes, so this is safe)
+                    ((c as u32).into(), c.len_utf8())
+                } else {
+                    // Not a valid UTF-8 character, use the first byte
+                    (bytes[0].into(), 1)
+                };
+                // Emit a warning if there are additional characters
+                if bytes.len() > len {
+                    show_warning!(
```

Comment:
I just realized you change the behaviour here, this used to return a `PartialMatch`. I'm not 100% sure if this is critical.

Reply:
(this makes me wonder if it's a good idea to move this code out of num_parser.rs:parse)

[-- sorry I need to run, I'll try to think more about this later]